pkg/kubectl/rolling_updater.go

```diff
@@ -188,37 +219,48 @@ func (r *RollingUpdater) Update(config *RollingUpdaterConfig) error {
 		return err
 	}
 	delete(newRc.ObjectMeta.Annotations, sourceIdAnnotation)
-	delete(newRc.ObjectMeta.Annotations, desiredReplicasAnnotation)
+	delete(newRc.ObjectMeta.Annotations, DesiredReplicasAnnotation)
 	newRc, err = r.updateAndWait(newRc, interval, timeout)
 	if err != nil {
 		return err
 	}
 
+	// Clean up the old RC based on policy.
 	switch config.CleanupPolicy {
 	case DeleteRollingUpdateCleanupPolicy:
 		// delete old rc
 		fmt.Fprintf(out, "Update succeeded. Deleting %s\n", oldName)
 		return r.c.DeleteReplicationController(r.ns, oldName)
 	case PreserveRollingUpdateCleanupPolicy:
+		fmt.Fprintf(out, "Update succeeded. Preserving %s\n", oldName)
 		return nil
 	default:
 		return nil
 	}
 }
 
-func (r *RollingUpdater) getExistingNewRc(sourceId, name string) (rc *api.ReplicationController, existing bool, err error) {
-	if rc, err = r.c.GetReplicationController(r.ns, name); err == nil {
-		existing = true
-		annotations := rc.ObjectMeta.Annotations
-		source := annotations[sourceIdAnnotation]
-		_, ok := annotations[desiredReplicasAnnotation]
-		if source != sourceId || !ok {
-			err = fmt.Errorf("Missing/unexpected annotations for controller %s: %s", name, annotations)
-		}
-		return
+func (r *RollingUpdater) getExistingNewRc(sourceId, name string) (*api.ReplicationController, bool, error) {
+	rc, err := r.c.GetReplicationController(r.ns, name)
+	if err != nil {
+		return nil, false, err
 	}
-	err = nil
-	return
+
+	if _, ok := rc.ObjectMeta.Annotations[DesiredReplicasAnnotation]; !ok {
+		err := fmt.Errorf("Missing %s annotation for controller %s", DesiredReplicasAnnotation, name)
+		return rc, true, err
+	}
+
+	// If there's no source associated with the RC, don't validate association.
+	if len(sourceId) == 0 {
```

Comment:
Nit: change to sourceId != 0 and move validation inside the block so you can reuse the existing successful return case (line 263).